oak_functions_service/src/lib.rs

```diff
@@ -58,22 +57,29 @@ pub trait Observer {
     fn wasm_invocation(&self, duration: core::time::Duration);
 }
 
-pub struct OakFunctionsService {
-    evidence: Evidence,
-    encryption_key_provider: Arc<EncryptionKeyProvider>,
+pub struct OakFunctionsService<
+    EKH: oak_restricted_kernel_sdk::EncryptionKeyHandle + 'static,
+    EP: oak_restricted_kernel_sdk::Evidencer,
```

Comment:
It's probably ok to accept a dynamic dispatch here (box + dyn) instead of dragging along the generic type bounds everywhere, but up to you (I mean, one of them is even within an Arc already, so an extra pointer indirection is completely negligible)

Reply:
Good point! I thought about that option also. 👍 i like this style personally, as it permits extra typesafety.

Reply:
eg the type signature for the testing setup function makes clear that it uses mocks, which I think is nice:

```
fn new_service_for_testing() -> OakFunctionsService<
    oak_restricted_kernel_sdk::MockEncryptionKeyHandle,
    oak_restricted_kernel_sdk::MockEvidenceProvidder,
>
```

Reply:
I see your point, but personally I still disagree.

My view is that it would be preferable that the same exact OakFunctionsService instance could be passed a different implementation of key handle / provider, and the choice of whether that is real of mock is concentrated in a single place (where that provider is created). (as it stands, the compiler would reify two distinct instances based on the concrete types). Having to repeat the fact that the key provider is the mock one everywhere does not improve things, it just makes it more annoying to make changes. From a type system point of view, the two (or more) implementations do the same thing, so it is acceptable that they should be interchangeable.

To me, the fact that the new_service_for_testing method returns the same exact type as the real (non testing) instance is actually a good thing (instead of a different type).

Anyways I just wanted to explain my point of view, I see yours, and I'll leave the choice to you.